improver/threshold.py

```diff
@@ -160,26 +152,39 @@ def __init__(
                 )
             fuzzy_factor_loc = fuzzy_factor
 
-        # Set fuzzy-bounds.  If neither fuzzy_factor nor fuzzy_bounds is set,
-        # both lower_thr and upper_thr default to the threshold value.  A test
-        # of this equality is used later to determine whether to process with
-        # a sharp threshold or fuzzy bounds.
         if fuzzy_bounds is None:
-            self.fuzzy_bounds = []
-            for thr in self.thresholds:
-                lower_thr = thr * fuzzy_factor_loc
-                upper_thr = thr * (2.0 - fuzzy_factor_loc)
-                if thr < 0:
-                    lower_thr, upper_thr = upper_thr, lower_thr
-                self.fuzzy_bounds.append((lower_thr, upper_thr))
+            self.fuzzy_bounds = self._generate_fuzzy_bounds(fuzzy_factor_loc)
         else:
-            self.fuzzy_bounds = fuzzy_bounds
+            self.fuzzy_bounds = (
+                [fuzzy_bounds] if isinstance(fuzzy_bounds, tuple) else fuzzy_bounds
```

Comment:
According to the docstring, `fuzzy_bounds` is a list of tuples. But here `fuzzy_bounds` can be a tuple. Maybe the docstring should be updated to reflect that. Perhaps this line could potentially be an issue if someone passed in a tuple of tuples.

Reply:
The type of `fuzzy_bounds` is envisaged to be consistent with `threshold` - that is if someone passes in a scalar threshold, they are expected to pass in a tuple of fuzzy bounds, and otherwise a list of tuples.  I'll admit this code is rather poor in that this is an implicit and unenforced expectation!  But that I believe was the thinking around how this would be used.

I think ideally I'd tidy this up to enforce the expectation.  However I think as a step towards improved code, this is better than it was, and there are currently other priorities on this end.  So I will update the docstring as you suggest.

Reply:
Great, thanks for making those changes. Everything looks good. And thanks for the explanation (especially as I'm new to the code base)!